src/source-odp.c

```diff
@@ -0,0 +1,300 @@
+ * \param pq pointer to the current PacketQueue
+ */
+TmEcode DecodeODP(ThreadVars *tv, Packet *p, void *data, PacketQueue *pq, PacketQueue *postpq)
+{
+    SCEnter();
+    DecodeThreadVars *dtv = (DecodeThreadVars *)data;
+
+    /* XXX HACK: flow timeout can call us for injected pseudo packets
+     *           see bug: https://redmine.openinfosecfoundation.org/issues/1107 
+     * Not implemented for ODP yet.
+     */
+    if (p->flags & PKT_PSEUDO_STREAM_END)
+        return TM_ECODE_OK;
+
+    /* update counters */
+    DecodeUpdatePacketCounters(tv, dtv, p);
+
+    /* call the decoder */
+    switch(p->datalink) {
+        case LINKTYPE_LINUX_SLL:
+            DecodeSll(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
+            break;
+        case LINKTYPE_ETHERNET:
+            DecodeEthernet(tv, dtv, p,GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
+            break;
+        case LINKTYPE_PPP:
+            DecodePPP(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
+            break;
+        case LINKTYPE_RAW:
+            DecodeRaw(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
+            break;
+        case LINKTYPE_NULL:
+            DecodeNull(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
+            break;
+        default:
+            SCLogError(SC_ERR_DATALINK_UNIMPLEMENTED, "Error: datalink type %" PRId32 " not yet supported in module DecodePcap", p->datalink);
+            break;
+    }
+
+    PacketDecodeFinalize(tv, dtv, p);
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode ReceiveODPBreakLoop(ThreadVars *ptv, void *arg)
+{
+    odp_loop_break = 1;
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/**
+ * \brief Registration Function for RecieveODP.
+ * \todo Unit tests are needed for this module.
+ */
+void TmModuleReceiveODPRegister (void)
+{
+    tmm_modules[TMM_RECEIVEPCAP].name = "ReceiveODP";
+    tmm_modules[TMM_RECEIVEPCAP].ThreadInit = ReceiveODPThreadInit;
+    tmm_modules[TMM_RECEIVEPCAP].Func = NULL;
+    tmm_modules[TMM_RECEIVEPCAP].PktAcqLoop = ReceiveODPLoop;
+    tmm_modules[TMM_RECEIVEPCAP].PktAcqBreakLoop = ReceiveODPBreakLoop;
+    tmm_modules[TMM_RECEIVEPCAP].ThreadExitPrintStats = NULL;
+    tmm_modules[TMM_RECEIVEPCAP].ThreadDeinit = NULL;
+    tmm_modules[TMM_RECEIVEPCAP].RegisterTests = NULL;
+    tmm_modules[TMM_RECEIVEPCAP].cap_flags = SC_CAP_NET_RAW;
+    tmm_modules[TMM_RECEIVEPCAP].flags = TM_FLAG_RECEIVE_TM;
+}
+
+static inline Packet *_odp_process_packet(ODPThreadVars *ptv, odp_packet_t opkt)
+{
+
+    uint32_t caplen = odp_packet_len(opkt);
+    u_char *pkt = odp_packet_data(opkt);
+    Packet *p;
+
+    p = (Packet *)odp_packet_push_tail(opkt, sizeof(Packet));
+    if (!p) {
+        fprintf(stderr, "error push_tail\n");
+        return NULL;
+    }
+
+    if (caplen == odp_packet_len(opkt)) {
+        fprintf(stderr, "packet not resized error push_tail\n");
+        return NULL;
+    }
+
+    /* @todo: remove memset, looks like it requires odp recompilation
```

Comment:
the packet structure has a mutex for certain corner cases, which is overwritten here. Some pthread implementations may not like this.